arch/arm/mach-zynq/platsmp.c

```diff
@@ -41,7 +41,7 @@ int zynq_cpun_start(u32 address, int cpu)
 
 	/* MS: Expectation that SLCR are directly map and accessible */
 	/* Not possible to jump to non aligned address */
-	if (!(address & 3) && (!address || (address >= trampoline_code_size))) {
+	if (!(address & 2) && (!address || (address >= trampoline_code_size))) {
```

Comment:
The part of this patch is in the tree already. It is not clear from your description why here should be 2 instead of 3 that's why please update.

Reply:
(address & 1) == 1  is not an unaligned address, it's just thumb code. Calling zynq_cpun_start(address_of_thumb2_function, 1) fails if you use & 3, which means the second CPU doesn't start with CONFIG_THUMB2_KERNEL enabled. With & 2 it works correctly.

Reply:
But that means that if we change the mask in the code from 3->2 then in normal mode address can end with 1 it will be executed which is also wrong. 
Xilinx is not enabling/testing thumb2 mode. I have built the kernel and SMP works and it looks like that all function start is aligned already. It means what's the problem which you are dealing with.